src/BalanceTx.purs

```diff
@@ -239,78 +244,58 @@ calculateMinFee' = calculateMinFee >>> map (rmap unwrap)
-        _, Nothing ->
-          pure $ Left $ UtxoIndexToUtxoError' CouldNotConvertUtxoIndex
-        Just utxos, Just utxoIndex'' -> do
-          let
-            -- Combines utxos at the user address and those from any scripts
-            -- involved with the contract in the unbalanced transaction. Don't
-            -- add collateral to this for now.
-            allUtxos :: Utxo
-            allUtxos = utxos `Map.union` utxoIndex''
-
-            -- After adding collateral, we need to balance the inputs and
-            -- non-Ada outputs before looping, i.e. we need to add input fees
-            -- for the Ada only collateral. No MinUtxos required. In fact perhaps
-            -- this step can be skipped and we can go straight to prebalancer.
-            unbalancedCollTx = addTxCollateral unbalancedTx collateral
-          -- Prebalance tx without fees:
-          case prebalanceCollateral zero allUtxos ownAddr unbalancedCollTx of
-            Left err -> pure $ Left err
-            Right ubcTx' -> do
-              fees' <- lmap CalculateMinFeeError' <$> calculateMinFee' ubcTx'
-              -- Prebalance tx with fees:
-              let
-                unbalancedCollTx' = fees' >>= \fees ->
-                  prebalanceCollateral
-                    (fees + feeBuffer)
-                    allUtxos
-                    ownAddr
-                    ubcTx'
-              case unbalancedCollTx' of
-                Left err -> pure $ Left err
-                Right unbalancedCollTx'' ->
-                  loop allUtxos ownAddr [] unbalancedCollTx'' >>= case _ of
-                    Left err -> pure $ Left err
-                    Right nonAdaBalancedCollTx ->
-                      lmap ReturnAdaChangeError' <$>
-                        returnAdaChange ownAddr allUtxos nonAdaBalancedCollTx
-                        >>= case _ of
-                          Left err -> pure $ Left err
-                          Right unsignedTx ->
-                            signTransaction unsignedTx
-                              <#> note (SignTxError' $ CouldNotSignTx ownAddr)
-
-  -- Some useful spies for debugging.
-  -- case signedTx' of
-  --   Left err -> pure $ Left err
-  --   Right signedTx -> do
-  --     let signedTxBody = (unwrap signedTx).body
-  --         signedTxOuputs = (unwrap signedTxBody).outputs
-  --         x = spy "signedTxInputVal" (getInputValue allUtxos signedTxBody)
-  --         y = spy "signedTxOutputs" signedTxOuputs
-  --         z = spy "signedTxOutputsVal" (Array.foldMap getAmount signedTxOuputs)
-  --         a = spy "signedTxFees" $ (unwrap signedTxBody).fee
-  --     pure $ pure $ signedTx
 
+-- | Balances an unbalanced transaction. For submitting a tx via Nami, the
+-- utxo set shouldn't include the collateral which is vital for balancing.
+-- In particular, the transaction inputs must not include the collateral.
+balanceTxM :: UnbalancedTx -> QueryM (Either BalanceTxError Transaction)
```

Comment:
Just a note: wouldn't it be a good idea to make transaction type wallet-specific?

Totally not a change request.

Reply:
This is a possibility and it depends how the project progresses and what wallets we allow. Eventually, we'll want multisig (which requires signing from different entities from potentially different wallets(?)) and this would still need to be able to balance a transaction for those various signatures.